remove prepare_payload (#2683)
ethereum/consensus-specs: specs/merge/validator.md

## Prerequisites

> This  documens is an extension of the [Altair -- Honest Validator](../altair/validator.md) guide.

Reviewer: accidentally changed?

Reviewer 2: I'm surprised that `codespell` allowed it. 🤔

| Name | SSZ equivalent | Description |
| - | - | - |
| `PayloadId` | `Bytes8` | Identifier of a payload building process |

## Helpers

> ### `get_pow_block_at_terminal_total_difficulty`

Author: Moved `get_pow_block_at_terminal_total_difficulty` and `get_terminal_pow_block` up to helpers section.

Not changes made to contents of the functions.

    if TERMINAL_BLOCK_HASH != Hash32():
        # Terminal block hash override takes precedence over terminal total difficulty
        pow_block_overrides = [block for block in pow_chain if block.block_hash == TERMINAL_BLOCK_HASH]
        if not any(pow_block_overrides):
            return None
        return pow_block_overrides[0]

    return get_pow_block_at_terminal_total_difficulty(pow_chain)
```

### `get_payload_id`

Given the `head_block_hash` and the `payload_attributes` that were used to
initiate the build process via `notify_forkchoice_updated`, `get_payload_id()`
returns the `payload_id` used to retrieve the payload via `get_payload`.

```python
def get_payload_id(parent_hash: Hash32, payload_attributes: PayloadAttributes) -> PayloadId:
    return PayloadId(
        hash(
            parent_hash
            + uint_to_bytes(payload_attributes.timestamp)
            + payload_attributes.random
            + payload_attributes.fee_recipient
        )[0:8]
    )
```

*Note*: This function does *not* use simple serialize `hash_tree_root` as to
avoid requiring simple serialize hashing capabilities in the Execution Layer.

## Protocols

### `ExecutionEngine`

> *Note*: `prepare_payload` and `get_payload` functions are added to the `ExecutionEngine` protocol for use as a validator.

Reviewer: ```suggestion
*Note*: `get_payload` function is added to the `ExecutionEngine` protocol for use as a validator.
```

> The body of each of these functions is implementation dependent.

Reviewer: ```suggestion
The body of this function is implementation dependent.
```

> The Engine API may be used to implement them with an external execution engine.

Reviewer: ```suggestion
The Engine API may be used to implement it with an external execution engine.
```

To obtain an execution payload, a block proposer building a block on top of a `state` must take the following actions:

1. Set `payload_id = prepare_execution_payload(state, pow_chain, finalized_block_hash, fee_recipient, execution_engine)`, where:
    * `state` is the state object after applying `process_slots(state, slot)` transition to the resulting state of the parent block processing
    * `pow_chain` is a list that abstractly represents all blocks in the PoW chain

>     * `finalized_block_hash` is the hash of the latest finalized execution payload (`Hash32()` if none yet finalized)

Reviewer: I would consider adding it to the state once we have a use case that can benefit from this addition. I am not sure that such use case already exists
